PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp

```diff
@@ -112,13 +150,24 @@ namespace // anonymous
 
 	generic_string getDialogFileName(IFileDialog* dialog)
 	{
+		generic_string fileName;
 		PWSTR pszFilePath = nullptr;
-		dialog->GetFileName(&pszFilePath);
-		generic_string fileName = pszFilePath;
+		HRESULT hr = dialog->GetFileName(&pszFilePath);
+		if (SUCCEEDED(hr) && pszFilePath)
+			fileName = pszFilePath;
 		CoTaskMemFree(pszFilePath);
 		return fileName;
 	}
 
+	generic_string getDialogFolder(IFileDialog* dialog)
+	{
+		com_ptr<IShellItem> psi;
+		HRESULT hr = dialog->GetFolder(&psi);
```

Comment:
`CoTaskMemFree` is not called for `psi`

Reply:
It is not needed here. We use `_com_ptr_t<T>` (`com_ptr` is just a shorthand alias).
It is just like `unique_ptr<T>` but works with COM interface `IUnknown`.
It calls `AddRef()` on copy and `Release()` in destructor.

There is the same solution for COM strings (that have to be freed with `CoTaskMemFree`) - `CComHeapPtr`. But I didn't want to add a dependency to ATL only because of it.

Reply:
About _com_ptr_t on [MSDN](https://docs.microsoft.com/en-us/cpp/cpp/com-ptr-t-class?view=msvc-160)